proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy

```diff
@@ -0,0 +1,727 @@
+  done
+
+lemma fault_handler_update_valid_cur_vcpu[wp]:
+  "option_update_thread thread (tcb_fault_handler_update \<circ> f) opt \<lbrace>valid_cur_vcpu\<rbrace>"
+  unfolding option_update_thread_def
+  by (wpsimp wp: thread_set_valid_cur_vcpu)
+
+lemma fault_handler_update_state_hyp_refs_of[wp]:
+  "option_update_thread thread (tcb_fault_handler_update \<circ> f) opt \<lbrace>\<lambda>s. P (state_hyp_refs_of s)\<rbrace>"
+  unfolding option_update_thread_def
+  by (fastforce intro: thread_set_hyp_refs_trivial split: option.splits)
+
+crunches set_mcpriority, set_priority
+  for valid_cur_vcpu[wp]: valid_cur_vcpu
+  (simp: set_priority_def)
+
+lemma ethread_set_state_hyp_refs_of[wp]:
+  "ethread_set f t \<lbrace>\<lambda>s. P (state_hyp_refs_of s)\<rbrace>"
+  apply (simp add: ethread_set_def set_eobject_def)
+  apply wp
+  apply (clarsimp dest!: get_tcb_SomeD)
+  done
+
+crunches tcb_sched_action, possible_switch_to, set_scheduler_action, reschedule_required
+  for state_hyp_refs_of[wp]: "\<lambda>s. P (state_hyp_refs_of s)"
+  (simp: tcb_sched_action_def set_tcb_queue_def get_tcb_queue_def possible_switch_to_def
+         reschedule_required_def set_scheduler_action_def)
+
+crunches set_mcpriority, set_priority
+  for state_hyp_refs_of[wp]: "\<lambda>s. P (state_hyp_refs_of s)"
+  (wp: thread_set_hyp_refs_trivial simp: set_priority_def thread_set_priority_def)
+
+lemma invoke_tcb_valid_cur_vcpu[wp]:
+  "\<lbrace>\<lambda>s. valid_cur_vcpu s \<and> sym_refs (state_hyp_refs_of s)\<rbrace>
+   invoke_tcb iv
+   \<lbrace>\<lambda>_ s. valid_cur_vcpu s\<rbrace>"
+  (is "\<lbrace>?pre\<rbrace> _ \<lbrace>_\<rbrace>")
+  apply (cases iv; clarsimp; (solves \<open>wpsimp wp: mapM_x_wp_inv\<close>)?)
+   defer
+   subgoal for tcb_ptr ntfn_ptr_opt
+     by (case_tac ntfn_ptr_opt; wpsimp)
+  \<comment> \<open>ThreadControl\<close>
+  apply (rename_tac target slot faultep mcp priority croot vroot buffer)
+  apply (rule validE_valid)
+  apply (rule hoare_seq_ext_skipE, solves wpsimp)+
+  apply (rule hoare_seq_ext_skipE)
+   apply (case_tac croot; clarsimp)
+    apply wpsimp
+   apply (rule_tac Q="\<lambda>_. ?pre" and E="\<lambda>_. ?pre" in hoare_post_impErr)
+     apply (rule hoare_seq_ext_skipE)
+      apply (wpsimp wp: cap_delete_valid_cur_vcpu)
+     apply (wpsimp wp: check_cap_inv)
+    apply simp
+   apply simp
+  apply (rule hoare_seq_ext_skipE)
+   apply (case_tac vroot; clarsimp)
+    apply wpsimp
+   apply (rule_tac Q="\<lambda>_. ?pre" and E="\<lambda>_. ?pre" in hoare_post_impErr)
+     apply (rule hoare_seq_ext_skipE)
+      apply (wpsimp wp: cap_delete_valid_cur_vcpu)
+     apply (wpsimp wp: check_cap_inv)
+    apply simp
+   apply simp
+  apply (rule hoare_seq_ext_skipE)
+   apply (case_tac buffer; clarsimp)
+    apply wpsimp
+   apply (rule hoare_seq_ext_skipE)
+    apply (wpsimp wp: cap_delete_valid_cur_vcpu)
+   apply (wpsimp wp: check_cap_inv hoare_drop_imps thread_set_hyp_refs_trivial thread_set_valid_cur_vcpu)
+  apply wpsimp
+  done
```

Comment:
This is one where forward stepping with invariant make sense, because with pure backwards wp reasoning would blow up over the case distinctions. No need to do all those case distinctions manually, though, `wpsimp` can do them on its own. In fact, you can nicely collapse the entire proof into:

```suggestion
  \<comment> \<open>ThreadControl\<close>
  apply (rule validE_valid)
  apply (rule hoare_seq_ext_skipE, solves ‹wpsimp wp: check_cap_inv›)+
  by (wpsimp wp: check_cap_inv hoare_drop_imps thread_set_hyp_refs_trivial thread_set_valid_cur_vcpu)
```

Reply:
Even for this, though, we should ideally use a proof method that does the skipping step, reliably handles special cases, and works for all variants of `valid`, `validE`, etc.

Not sure it makes sense to put that into this PR, but it's not good to let the assembly-level proof steps proliferate too much, because they'll be hard to remove later if we wait too long.